### ISSUE TITLE
rss2email issue regarding XDG_RUNTIME_DIR and lockfile

### ISSUE THREAD

**Issue by Ekleog** [rss2email/rss2email#136]
(Here is a copy of an email, from an email from `Lucas <lucas@sexy.is>` who doesn't have a github account. I haven't reviewed it in-depth, and it includes a patch that should solve the issue raised)

Hey,

As we talked on IRC, attached you can find a reproduction script for
the issue.

To recap:
- rss2email/main.py:172 creates a lockfile
- Base directory for the lockfile is XDG_RUNTIME_DIR, otherwise /tmp
- lockfile name is the same for every user
- As long as users don't set XDG_RUNTIME_DIR, lockfile is always
  /tmp/rss2email.lock, making it impossible to use `r2e` after another
  user used it

Possible solutions / workarounds:
1. Use a personalized, per user directory when XDG_RUNTIME_DIR isn't
   set: for example, tmux uses `/tmp/tmux-$uid` with permissions `0700`.
2. Tell the user to always use XDG_RUNTIME_DIR

For 1., check the attached patch. After extracting it, you can apply it
running, inside the repository:

    git am 0001-Create-lockfile-in-an-user-specific-directory.patch

That will keep my authorship in place. It's a PR in mail, basically :P
Patch is made against commit 34c924173a8858e3dd8aa914d53f77e3e43fefc8,
current master as of time of writing. If you need to create a PR
anyways from it, you can apply the patch and push against a branch, and
create the PR from that branch.

Cheers!
-Lucas

`[ 0001-Create-lockfile-in-an-user-specific-directory.patch: text/plain ]`
```diff
From 80c29040fecc40871f460b972f59b1508e218374 Mon Sep 17 00:00:00 2001
Message-Id: <80c29040fecc40871f460b972f59b1508e218374.1599017054.git.lucas@sexy.is>
From: Lucas <lucas@sexy.is>
Date: Wed, 2 Sep 2020 03:15:25 +0000
Subject: [PATCH] Create lockfile in an user-specific directory

Current approach makes `r2e` create the lockfile under
`/tmp/rss2email.lock` if XDG_RUNTIME_DIR is unset, for all users,
making it impossible to use in shared shell environments. Instead,
create the lockfile under `/tmp/rss2email-${UID}/rss2email.lock`.

While there, document that XDG_RUNTIME_DIR is used for storing the
lockfile.
---
 r2e.1             | 4 ++++
 rss2email/main.py | 7 ++++++-
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/r2e.1 b/r2e.1
index cc50a50..e329eb0 100644
--- a/r2e.1
+++ b/r2e.1
@@ -302,6 +302,10 @@ $HOME/\&.config.
 The preferred directory for data files.  Defaults to
 $HOME/\&.local/share.
 .TP
+.B XDG_RUNTIME_DIR
+The preferred directory for lockfiles.  Defaults to
+/tmp/rss2email-UID
+.TP
 .B XDG_CONFIG_DIRS
 A colon ':' separated, preference ordered list of base directories for
 configuration files in addition to $XDG_CONFIG_HOME.  Defaults to
diff --git a/rss2email/main.py b/rss2email/main.py
index 3581ceb..7e243bc 100644
--- a/rss2email/main.py
+++ b/rss2email/main.py
@@ -169,7 +169,12 @@ def run(*args, **kwargs):
     # Immediately lock so only one r2e instance runs at a time
     if UNIX:
         import fcntl as _fcntl
-        lockfile_path = _os.path.join(_os.environ.get("XDG_RUNTIME_DIR", default="/tmp/"), "rss2email.lock")
+        from pathlib import Path as _Path
+        dir = _os.environ.get("XDG_RUNTIME_DIR")
+        if dir is None:
+            dir = _os.path.join("/tmp", "rss2email-{}".format(_os.getuid()))
+            Path(dir).mkdir(mode=0o700, parents=True, exists_ok=True)
+        lockfile_path = _os.path.join(dir, "rss2email.lock")
         lockfile = open(lockfile_path, "w")
         _fcntl.lockf(lockfile, _fcntl.LOCK_EX)
         _LOG.debug("acquired lock file {}".format(lockfile_path))
-- 
2.28.0
```

`[ rss2email-repro.sh: text/plain ]`
```sh
#!/bin/sh
set -x

tmpd=$(mktemp -d) || { echo "can't create temp dir" >&2; exit 1; }
trap 'rm -fr "$tmpd"' EXIT INT QUIT TERM

XDG_CONFIG_HOME=$tmpd
XDG_DATA_HOME=$tmpd
export XDG_CONFIG_HOME XDG_DATA_HOME

r2e new email@domain.invalid
r2e add rss2email https://github.com/rss2email/rss2email/releases.atom

# Ensure there is no rss2email lock file
# DON'T RUN WHILE RUNNING OTHER rss2email INSTANCE
rm -f /tmp/rss2email.lock

# sudo is needed to create the file with another user. It isn't
# necessary to use the root user--any different from the current user
# will do
sudo env XDG_CONFIG_HOME="$tmpd" XDG_DATA_HOME="$tmpd" r2e run --no-send

# This will show the file is owned by a user different than current one
ls -l /tmp/rss2email.lock

# This will fail this time, getting a permission denied in the lock file
r2e run --no-send
```

**Comment by Ekleog** [comment]
This needs:
- review
- integration of the reproducer script in the CI
- landing the reproducer script in the CI and the patch fix

**Comment by auouymous** [comment]
Should the import be moved to top of file?

Path is imported as _Path and then used as `Path(dir)` which fails.

An invalid `exists_ok` parameter is passed to mkdir which fails, should be `exist_ok`.

It looks good and works fine after fixing those two issues.

**Comment by Ekleog** [comment]
From `Lucas <lucas@sexy.is>` (sorry for the delay on my side copying it here, I've been a bit overwhelmed this week):

Hey,

I addressed the comment, and this time I did the things well and
tested the patch before sending. All 23 tests completed successfully.

Updated patch is attached, rebased onto master. As for integrating the
reproducer into CI, amn't sure if it's worth the trouble, nor if GitHub
Actions are flexible enough to run something under two different users.

BTW, thanks a lot for taking the time of receiving the reports and
patches over email. I really appreciate it.

Cheers,
-Lucas

```diff
From cce15dce7c897685e69962aa10c101692a3c1ed7 Mon Sep 17 00:00:00 2001
Message-Id: <cce15dce7c897685e69962aa10c101692a3c1ed7.1599337080.git.lucas@domain.invalid>
From: Lucas <lucas@sexy.is>
Date: Wed, 2 Sep 2020 03:15:25 +0000
Subject: [PATCH] Create lockfile in an user-specific directory

Current approach makes `r2e` create the lockfile under
`/tmp/rss2email.lock` if XDG_RUNTIME_DIR is unset, for all users,
making it impossible to use in shared shell environments. Instead,
create the lockfile under `/tmp/rss2email-${UID}/rss2email.lock`.

While there, document that XDG_RUNTIME_DIR is used for storing the
lockfile.
---
 r2e.1             | 4 ++++
 rss2email/main.py | 7 ++++++-
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/r2e.1 b/r2e.1
index cc50a50..e329eb0 100644
--- a/r2e.1
+++ b/r2e.1
@@ -302,6 +302,10 @@ $HOME/\&.config.
 The preferred directory for data files.  Defaults to
 $HOME/\&.local/share.
 .TP
+.B XDG_RUNTIME_DIR
+The preferred directory for lockfiles.  Defaults to
+/tmp/rss2email-UID
+.TP
 .B XDG_CONFIG_DIRS
 A colon ':' separated, preference ordered list of base directories for
 configuration files in addition to $XDG_CONFIG_HOME.  Defaults to
diff --git a/rss2email/main.py b/rss2email/main.py
index 3581ceb..4b30478 100644
--- a/rss2email/main.py
+++ b/rss2email/main.py
@@ -169,7 +169,12 @@ def run(*args, **kwargs):
     # Immediately lock so only one r2e instance runs at a time
     if UNIX:
         import fcntl as _fcntl
-        lockfile_path = _os.path.join(_os.environ.get("XDG_RUNTIME_DIR", default="/tmp/"), "rss2email.lock")
+        from pathlib import Path as _Path
+        dir = _os.environ.get("XDG_RUNTIME_DIR")
+        if dir is None:
+            dir = _os.path.join("/tmp", "rss2email-{}".format(_os.getuid()))
+            _Path(dir).mkdir(mode=0o700, parents=True, exist_ok=True)
+        lockfile_path = _os.path.join(dir, "rss2email.lock")
         lockfile = open(lockfile_path, "w")
         _fcntl.lockf(lockfile, _fcntl.LOCK_EX)
         _LOG.debug("acquired lock file {}".format(lockfile_path))
```

**Comment by Ekleog** [comment]
This LGTM to me to land (through a PR so that CI can run), does anyone dissent?

**Comment by Ekleog** [comment]
(also, thank you @auouymous for the review!)

**Comment by auouymous** [comment]
LGTM

**Comment by Ekleog** [comment]
This already landed in #142 :)
